exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,2155 @@
+		instrumentType = strings.ToUpper(ok.GetInstrumentTypeFromAssetItem(assetType))
+		if !(instrumentType == okxInstTypeSpot ||
+			instrumentType == okxInstTypeMargin ||
+			instrumentType == okxInstTypeSwap ||
+			instrumentType == okxInstTypeFutures ||
+			instrumentType == okxInstTypeOption) {
+			instrumentType = okxInstTypeANY
+		}
+	}
+	if len(tooglers) > 2 && tooglers[2] {
+		if !pair.IsEmpty() {
+			underlying, _ = ok.GetUnderlying(pair, assetType)
+		}
+	}
+	if len(tooglers) > 1 && tooglers[1] {
+		format, err = ok.GetPairFormat(assetType, false)
+		if err != nil {
+			return nil, err
+		}
+		if pair.Base.String() == "" || pair.Quote.String() == "" {
+			return nil, errIncompleteCurrencyPair
+		}
+		instrumentID = format.Format(pair)
+		if err != nil {
+			instrumentID = ""
+		}
+	}
+	if channel == "" {
+		return nil, errMissingValidChannelInformation
+	}
+	input := &SubscriptionOperationInput{
+		Operation: operation,
+		Arguments: []SubscriptionInfo{
+			{
+				Channel:        channel,
+				InstrumentType: instrumentType,
+				Underlying:     underlying,
+				InstrumentID:   instrumentID,
+			},
+		},
+	}
+	ok.WsRequestSemaphore <- 1
+	defer func() { <-ok.WsRequestSemaphore }()
+	err = ok.Websocket.Conn.SendJSONMessage(input)
+	if err != nil {
+		return nil, err
+	}
+	timer := time.NewTimer(ok.WebsocketResponseMaxLimit)
+	wsResponse := make(chan *wsIncomingData)
+	randomID, err := common.GenerateRandomString(16)
+	if err != nil {
+		return nil, fmt.Errorf("%w, generating random string for incoming websocket response failed", err)
+	}
+	ok.WsResponseMultiplexer.Register <- &wsRequestInfo{
+		ID:             randomID,
+		Chan:           wsResponse,
+		Event:          input.Operation,
+		Channel:        input.Arguments[0].Channel,
+		InstrumentType: input.Arguments[0].InstrumentType,
+		InstrumentID:   input.Arguments[0].InstrumentID,
+	}
+	defer func() { ok.WsResponseMultiplexer.Unregister <- randomID }()
+	for {
+		select {
+		case data := <-wsResponse:
+			if strings.EqualFold(data.Event, operation) &&
+				data.Argument.Channel == input.Arguments[0].Channel &&
+				data.Argument.InstrumentType == input.Arguments[0].InstrumentType &&
+				data.Argument.InstrumentID == input.Arguments[0].InstrumentID {
+				return data.copyToSubscriptionResponse(), nil
+			}
+			continue
+		case <-timer.C:
+			timer.Stop()
+			return nil, fmt.Errorf("%s websocket connection: timeout waiting for response with an operation: %v",
+				ok.Name,
+				input.Operation)
+		}
+	}
+}
+
+// Private Channel Websocket methods
+
+// WsAuthChannelSubscription send a subscription or unsubscription request for different channels through the websocket stream.
+func (ok *Okx) WsAuthChannelSubscription(operation, channel string, assetType asset.Item, pair currency.Pair, uid, algoID string, tooglers ...bool) (*SubscriptionOperationResponse, error) {
+	if !(strings.EqualFold(operation, "subscribe") || strings.EqualFold(operation, "unsubscribe")) {
```

Comment:
```suggestion
	if !strings.EqualFold(operation, "subscribe") && !strings.EqualFold(operation, "unsubscribe") {
```
Does this need to be equalfold?